ENH: support non default indexes in writing to Parquet

dhirschfeld opened the pull request:
- [x] closes #18581
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

jorisvandenbossche commented:
We need to check how this interacts with the `columns` keyword when reading back in. Ideally selecting only a few columns to read will still preserve your index (I think?), but I guess this is currently not happening (at least with pyarrow).

jschendel commented on pandas/io/parquet.py:
```diff
@@ -147,27 +148,30 @@ def to_parquet(df, path, engine='auto', compression='snappy', **kwargs):
+                             "to make the index into column(s)".format(
+                                 type(df.index)))
+
+        if not df.index.equals(RangeIndex.from_range(range(len(df)))):
```
Could create the `RangeIndex` via `RangeIndex(len(df))` here to make this a little cleaner.

dhirschfeld replied:
Done.

jreback commented on pandas/io/parquet.py:
```diff
@@ -147,27 +148,30 @@ def to_parquet(df, path, engine='auto', compression='snappy', **kwargs):
-    if df.index.name is not None:
-        raise ValueError("parquet does not serialize index meta-data on a "
-                         "default index")
+    # *unless* we're using pyarrow >= 0.7.1 which does support multi-indexes
```
so it *might* be better (though a bit more refactoring), to move the validation of index and validation of columns into methods on a base class to the impl

e.g.

```
class BaseImpl(object):


    def validate_index(....):
       ..

    def validate_columns(....):
      ....


class PyArrowImpl(BaseImpl):

     def write(...):
          if pyarrow < 0.7.1:
             self.validate_index()

         self.validate_columns()

class FastParquetImpl(BaseImpl):
      .....
```

as it makes things cleaner.

dhirschfeld replied:
Refactored in the latest commits...

jreback commented on pandas/tests/io/test_parquet.py:
```diff
@@ -360,6 +333,61 @@ def test_categorical_unsupported(self, pa_lt_070):
+            df.index = index
+            self.check_round_trip(df, engine)
+
+        # index with meta-data
```
could factor these 2 tests here from pyarrow impl (and also from fastparquet) into another test (that succeeds on all impls and versions)

dhirschfeld replied:
Since the behaviour changes based on the library and the version of the library I'm not sure how else to do it while maintaining test coverage on all versions.

Feel free to push a fix if you'd like as I'm knocking off for the evening shortly...

jreback reviewed (changes requested):
lgtm. some changes; pls. add a bug fix note in 0.22.0

jreback commented on pandas/io/parquet.py:
```diff
@@ -34,7 +34,50 @@ def get_engine(engine):
+        self._validate_index(df)
+
+    def write(self, df, path, compression, **kwargs):
+        raise NotImplementedError()
```
import and use ``pandas.core.common.AbstractMethodError`` here instead

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -77,6 +77,7 @@ Other Enhancements
 - :func:`Series.fillna` now accepts a Series or a dict as a ``value`` for a categorical dtype (:issue:`17033`)
 - :func:`pandas.read_clipboard` updated to use qtpy, falling back to PyQt5 and then PyQt4, adding compatibility with Python3 and multiple python-qt bindings (:issue:`17722`)
 - Improved wording of ``ValueError`` raised in :func:`read_csv` when the ``usecols`` argument cannot match all columns. (:issue:`17301`)
+- Enabled the use of non-default indexes in ``to_parquet`` with pyarrow>=0.7.0 (:issue:`18581`)
```
use ``:func:`DataFrame.to_parquet` `` here, use double-backticks around ``pyarrow>=0.7.0``

jreback reviewed (changes requested):
minor comments. if you can refactor the tests a bit ok (if not no big deal). ping on green.

jorisvandenbossche reviewed (changes requested):
`fastparquet` also supports writing the index, so no need to still raise the errors in that case.

Also, to repeat my comment from before: we need to decide how this interacts with `columns` keyword before merging this.

dhirschfeld commented:
> *`fastparquet` also supports writing the index, so no need to still raise the errors in that case.*

It seems it might just be easier all-round to just bump the pyarrow/fastparquet deps to the min required to support non-default indices - i.e. `>=0.7.0`  for pyarrow and ??? for fastparquet

jreback commented:
would be ok with bumping pyarrow dep
i think fp is ok as is (i bumped it when doing this originally)

pep8speaks commented:
Hello @dhirschfeld! Thanks for updating the PR.

 - In the file [`pandas/io/parquet.py`](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/io/parquet.py), following are the PEP8 issues :

> [Line 66:1](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/io/parquet.py#L66): [E112](https://duckduckgo.com/?q=pep8%20E112) expected an indented block
> [Line 66:1](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/io/parquet.py#L66): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 0
> [Line 66:3](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/io/parquet.py#L66): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 66:7](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/io/parquet.py#L66): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator

 - In the file [`pandas/tests/io/test_parquet.py`](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py), following are the PEP8 issues :

> [Line 46:1](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L46): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 0
> [Line 46:3](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L46): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 46:7](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L46): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 64:1](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L64): [E112](https://duckduckgo.com/?q=pep8%20E112) expected an indented block
> [Line 64:1](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L64): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 0
> [Line 64:3](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L64): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 64:5](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L64): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 64:7](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L64): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 65:5](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L65): [E113](https://duckduckgo.com/?q=pep8%20E113) unexpected indentation
> [Line 66:1](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L66): [E112](https://duckduckgo.com/?q=pep8%20E112) expected an indented block
> [Line 66:7](https://github.com/pandas-dev/pandas/blob/44b0fe8231bbc25b4b18e368b90cce539a608b83/pandas/tests/io/test_parquet.py#L66): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator

jreback commented on ci/requirements-3.5.sh:
```diff
@@ -8,4 +8,4 @@ echo "install 35"
 conda remove -n pandas python-dateutil --force
 pip install python-dateutil
 
```
can you check if we have pyarrow installed in ci/* anywhere else (and if so update the pinned versions)

dhirschfeld replied:
I checked and changed any pinned versions to the new minimum. If it wasn't already pinned I left it as is since a normal install will pull in later versions that the minimum.

dhirschfeld replied:
![image](https://user-images.githubusercontent.com/881019/33705822-888b68dc-db7d-11e7-86db-f429e370406c.png)

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -163,7 +163,7 @@ Other Enhancements
 - Improved wording of ``ValueError`` raised in :func:`read_csv` when the ``usecols`` argument cannot match all columns. (:issue:`17301`)
 - :func:`DataFrame.corrwith` now silently drops non-numeric columns when passed a Series. Before, an exception was raised (:issue:`18570`).
-
+- Enabled the use of non-default indexes in :func:`DataFrame.to_parquet` where the underlying engine supports it (:issue:`18581`)
```
also there is another section on changed deps, pls add another entry which says pyarrow is now min 0.7.0 for parquet (also may need to update in io.rst)

jreback reviewed (changes requested):
doc changes, ping on green.

jorisvandenbossche commented:
@jreback not merge on green, see my request for changes on which I got no reply yet

BTW, I think we can do this for the 0.21.1 milestone. The other enhancements to `to/read_parquet` also went into that one (seeing it as fixes on very new functionality)

dhirschfeld commented:
@jorisvandenbossche - as of `fastparquet=0.1.3` it doesn't appear to support multi-indexes

![image](https://user-images.githubusercontent.com/881019/33709251-7f1d50f2-db88-11e7-98ef-f3700ca48b2c.png)


...at least reading them anyway. Interestingly, it seems to write them fine and the file written by fastparquet can then be read in perfectly by pyarrow.

Not sure what to do about that but since fastparquet throws a sensible error I'd be inclined to just leave it and a future update will likely add the missing functionality.

dhirschfeld commented:
@jorisvandenbossche - you're right about the `columns` argument not preserving the index:

![image](https://user-images.githubusercontent.com/881019/33709668-c3f41fa2-db89-11e7-9d8f-628313b0a138.png)

The workaround, at the cost of efficiency, is simple enough:
```python
pd.io.parquet.read_parquet('tmp_pa.pq', engine='pyarrow')[['A','B']]
```
...but you have to know beforehand that the file was saved with a multi-index - i.e. it's a silent failure mode which isn't ideal.

If I name the index levels you can actually pull them out by name:
```python
df = pd.DataFrame(randn(2*len(dates), 3), columns=list('ABC'))
index = pd.MultiIndex.from_product([['Level1', 'Level2'], dates], names=['level','date'])
df.index = index
```
![image](https://user-images.githubusercontent.com/881019/33710127-5591168a-db8b-11e7-9b9c-868bd4494e24.png)
```python
pd.io.parquet.read_parquet('tmp_pa.pq', engine='pyarrow', columns=['A','B','level','date']).head()
```
![image](https://user-images.githubusercontent.com/881019/33710207-92ebbd96-db8b-11e7-82e0-886d1286afe8.png)

...so, ideally if pandas can tell if the file has a multi-index the index names could then always be prepended to the columns argument.

dhirschfeld commented:
@jorisvandenbossche - were there other changes you wanted that I've missed?

jorisvandenbossche commented:
> Not sure what to do about that but since fastparquet throws a sensible error I'd be inclined to just leave it and a future update will likely add the missing functionality.

It's a bit unfortunate it only comes up when reading it back in, so either we are ok with that, or either we need to to raise on writing MIs with fastparquet.

> If I name the index levels you can actually pull them out by name:

That will not always work. Eg if they have no names, and pyarrow master has also changed this and the index levels will never have this name, so the workaround you show will not work there. 

The thing we can do is the following:

```
In [85]: import pyarrow.parquet as pq

In [86]: f = pq.ParquetFile("__test_mi.parquet")

In [87]: f.metadata.metadata[b'pandas']
Out[87]: b'{"columns": [{"name": "A", "pandas_type": "float64", "numpy_type": "float64", "metadata": null}, {"name": "B", "pandas_type": "float64", "numpy_type": "float64", "metadata": null}, {"name": "C", "pandas_type": "float64", "numpy_type": "float64", "metadata": null}, {"name": "__index_level_0__", "pandas_type": "unicode", "numpy_type": "object", "metadata": null}, {"name": "__index_level_1__", "pandas_type": "datetime", "numpy_type": "datetime64[ns]", "metadata": null}], "index_columns": ["__index_level_0__", "__index_level_1__"], "pandas_version": "0.22.0.dev0+288.g6b6cfb8"}'

In [88]: import json

In [89]: pandas_metadata = json.loads(f.metadata.metadata[b'pandas'].decode('utf8'))

In [90]: pandas_metadata['index_columns']
Out[90]: ['__index_level_0__', '__index_level_1__']
```

and then add those to the `columns` list. 
And something similar for `fastparquet`:

```
In [98]: import fastparquet

In [99]: f = fastparquet.ParquetFile("__test_mi.parquet")

In [100]: json.loads(f.key_value_metadata['pandas'])['index_columns']
Out[100]: ['__index_level_0__', '__index_level_1__']
```

dhirschfeld commented:
Heh, beat me to it:
```python
import pyarrow.parquet as pq
filemetadata = pq.read_metadata('tmp_fp.pq')
metadata = eval(filemetadata.metadata.get(b'pandas'), None, dict(null=None))
metadata['index_columns']
```
...eval is pretty nasty!

jorisvandenbossche commented:
I also think that *if* there is a default index (range-like index without name, the case that was allowed now), we should not write it to the parquet file, because that would just be waste of time and space.

jorisvandenbossche commented:
Yep, the `json.loads` is the better alternative :-)

jreback commented:
> BTW, I think we can do this for the 0.21.1 milestone. The other enhancements to to/read_parquet also went into that one (seeing it as fixes on very new functionality)

since we are bumping the version I don't think this is a good idea. marked for 0.22

I am ok with this fix, @jorisvandenbossche

dhirschfeld commented:
For pyarrow the columns fix is as simple as setting `use_pandas_metadata=True`:
```python
    def read(self, path, columns=None, **kwargs):
        path, _, _ = get_filepath_or_buffer(path)
        parquet_file = self.api.parquet.ParquetFile(path)
        kwargs['use_pandas_metadata'] = True
        return parquet_file.read(columns=columns, **kwargs).to_pandas()
```

For fastparquet I would use:
```python
    def read(self, path, columns=None, **kwargs):
        path, _, _ = get_filepath_or_buffer(path)
        parquet_file = self.api.ParquetFile(path)
        metadata = json.loads(parquet_file.key_value_metadata['pandas'])
        if columns is not None:
            columns = set(chain(columns, metadata['index_columns']))
        return parquet_file.to_pandas(columns=columns, **kwargs)
```
...but since it doesnt' support multi-indexes anyway it's a moot point.

jreback commented:
@martindurant can you weigh in here on the fp fix?

jorisvandenbossche commented:
> ...but since it doesnt' support multi-indexes anyway it's a moot point.

It's not a moot point, as it still supports single level index, which is better than nothing?

The fastparquet fix is fine, the MultiIndex 'not implemented' error is just an enhancement that would be welcome in fastparquet.

jorisvandenbossche commented:
> I am ok with this fix, @jorisvandenbossche

I am not yet fine with the current version, so please don't merge.

> since we are bumping the version I don't think this is a good idea. marked for 0.22

If we do that, it's 0.21.1 that will become 0.22 and 0.22 will become 0.23 no?

dhirschfeld commented:
I've pushed some changes to handle the columns issue - at least with pyarrow. The test is marked as xgail for fastparquet but that can be easily changed if fastparquet adds support for multi-indexes.

I think I've covered everything but let me know if there are further changes you'd like...

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -34,82 +35,95 @@ def get_engine(engine):
+    @staticmethod
+    def validate_dataframe(df):
+        if not isinstance(df, DataFrame):
+            raise ValueError("to_parquet only support IO with DataFrames")
```
support -> supports

jreback commented:
> if we do that, it's 0.21.1 that will become 0.22 and 0.22 will become 0.23 no?

huh?

no 0.21.1 - should release this soon as a bugfix, 0.22 for major changes.

dhirschfeld commented on pandas/io/parquet.py:
```diff
@@ -120,7 +134,8 @@ def write(self, df, path, compression='snappy', **kwargs):
         path, _, _ = get_filepath_or_buffer(path)
-        return self.api.ParquetFile(path).to_pandas(columns=columns, **kwargs)
+        parquet_file = self.api.ParquetFile(path)
+        return parquet_file.to_pandas(columns=columns, **kwargs)
```
For fastparquet I haven't bothered reading the metadata to pull out the columns since even if you do you still hit the `NotImplementedError`. I figure if fastparquet do add support for multi-indexes they might implement a solution similar to pyarrow so that the user doesn't have to introspect the metadata to reconstruct the correct DataFrame.

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -34,82 +35,95 @@ def get_engine(engine):
+            raise ValueError("to_parquet only support IO with DataFrames")
+        # must have value column names (strings only)
+        if df.columns.inferred_type not in {'string', 'unicode'}:
+            raise ValueError("parquet must have string column names")
```
We should also test the index level names to be strings (if not None, and if written)

jorisvandenbossche commented:
@dhirschfeld Thanks for the updates! Will review in a moment

jorisvandenbossche commented:
> For fastparquet I haven't bothered reading the metadata to pull out the index columns 

Apparently it is doing this 'correctly' by default (`f` is a fastparquet ParquetFile):

```
In [142]: f.to_pandas()
Out[142]: 
    bar  baz  foo
zoo              
x     A    1  one
y     B    2  one
z     C    3  one
q     A    4  two
w     B    5  two
t     C    6  two

In [143]: f.to_pandas(columns=['bar'])
Out[143]: 
    bar
zoo    
x     A
y     B
z     C
q     A
w     B
t     C
```

So also there this "extracting metadata" is not needed

jorisvandenbossche commented on pandas/tests/io/test_parquet.py:
```diff
@@ -262,40 +269,34 @@ def test_write_with_index(self, engine):
+            df.to_parquet(path, engine)
+            result = read_parquet(path, engine)
+            expected = df
+            tm.assert_frame_equal(result, expected)
```
Is there a reason you can't use `check_round_trip` here?

jorisvandenbossche commented on pandas/tests/io/test_parquet.py:
```diff
@@ -262,40 +269,34 @@ def test_write_with_index(self, engine):
+        df = pd.DataFrame(randn(2*len(dates), 3), columns=list('ABC'))
+        index = pd.MultiIndex.from_product(
+            [['Level1', 'Level2'], dates],
+            names=['level', 'date']
```
Can you do the same test without names? (maybe parameterize)

dhirschfeld commented:
Yep, fastparquet works fine with a non-multi-index DataFrame. The code I posted above would *also* work but was moot in the sense that it didn't fix the multi-index problem.

jorisvandenbossche commented:
Ah, OK, then I understand your "moot" point :-)

jorisvandenbossche reviewed (commented):
Looking good:

- Can you update the io.rst docs about parquet and index ?

dhirschfeld commented:
Thanks for the reviews @jreback, @jorisvandenbossche & @jschendel!

Sorry it's taken a while to get here - I've usually only got a small window in the evenings when I can work on stuff not directly related to my *actual* work. That window is fast closing down this way so feel to push any changes in the interim. I'm also happy to pick it up again tomorrow though. No big rush if it's scheduled for 0.22...

jorisvandenbossche commented:
Maybe not essential for this PR, but I think we should also think about:

- not writing the index if it is a default index (range-like index without name). It will be cheaper to just reconstruct it on reading (which is already done by both engines if there is no index column in the metadata)

- providing a `index=True/False` keyword similar to `to_csv` and others to control if you want to write the index or not (both pyarrow and fastparquet support this)

jorisvandenbossche commented:
> Sorry it's taken a while to get here - I've usually only got a small window in the evenings when I can work on stuff not directly related to my actual work.

No problem! The PR is only open yet for 3 days :-) And it will be a very nice improvement.

jorisvandenbossche commented:
@jreback to come back on the milestone. I personally would like to see this released as soon as possible, as it is an important change to make `read_parquet`/`to_parquet` as brand new functionality easier to use. This will impact how users use those functions, so let's change that as quick as possible. Since we are also including the addition of the `columns` keyword in the 0.21.1 release, I don't see why we wouldn't include this in the 0.21.1 release. 
(I can help with finishing this PR if time is a problem)

> since we are bumping the version I don't think this is a good idea. 

I don't fully understand what bumping the version has to do with it, but let's keep the discussion about version numbers for https://github.com/pandas-dev/pandas/issues/18244, and let's assume here that we want to release 0.21.1 in a couple of days (but as argued above, I would like to try to include it in that release).

jreback commented:
well, almost all of the 0.21.1 fixes won't break your existing environment; that's the point.

This is actually something that will. that's why its 0.22.0. its an api breaking change. Now personally I agree that it would be fine to break this as its a new feature. If you want to be less strict on this because of that, then I am not opposed.

jorisvandenbossche commented:
This will not really break code, it just enables a case that previously gave an error and people had to do workarounds (resetting the index), and thus will prevent that a lot of people that try this new functionality out have to do workarounds that will soon not be needed any more.

I know that I typically want to be rather strict on what gets included in a bug-fix release, but since this is brand new functionality + the other improvements to it already went in, I am fine with it in this case

jreback commented:
no it *will* break code. if you have pyarrow < 0.7 installed, you are out of luck

jreback commented:
> I know that I typically want to be rather strict on what gets included in a bug-fix release, but since this is brand new functionality + the other improvements to it already went in, I am fine with it in this case

what other improvements?

jorisvandenbossche commented:
> no it will break code. if you have pyarrow < 0.7 installed, you are out of luck

ok, yes, for that part of the PR it is indeed breaking. But for the actual enhancement of enabling indexes (if you already had the latest pyarrow release) not.

> what other improvements?

The `columns` one and the "passing kwargs to engine" one.

jorisvandenbossche commented:
> no it will break code. if you have pyarrow < 0.7 installed, you are out of luck

we could also split this PR -> add the ability to write index if pyarrow >= 0.7, and keep dropping support for arrow < 0.7 for a separate PR.

dhirschfeld commented:
It would be possible to make it only fail in the read/write methods *if* there was a non-default index and pyarrow<0.7, rather than failing immediately in the constructor. 

Does that sound like a good compromise? If so I can take a crack at it today...

jorisvandenbossche commented:
Yes, that would be a good compromise I think

dhirschfeld commented:
It appears in `pyarrow=0.4.1` *reading* a multi-index DataFrame works
![image](https://user-images.githubusercontent.com/881019/33792890-f435680a-dcf7-11e7-856f-cdb658b28a71.png)
...though not columns, without the metadata introspection trick.

dhirschfeld committed:
Allow non-default indexes in to_parquet.

...when supported by the underlying engine.
Fixes #18581

dhirschfeld commented:
Had to squash the commits to get out of rebase hell!

I think I've addressed all comments and concerns so would be keen to get this in. As it is now it will still support pyarrow >= 0.4.1 up to the extent the library itself supports reading/writing parquet. 

TODO: Re-add the whatsnew entries. I wasn't sure if they should now go in v0.21.1 or not?

If there are other changes you'd like let me know...

dhirschfeld committed:
Fixed test_write_with_index on Windows

snappy compression isn't available on Windows

dhirschfeld commented:
I think the appveyor error is legit.
```
>       raise AssertionError(msg)
E       AssertionError: DataFrame.index are different
E       
E       Attribute "names" are different
E       [left]:  ['index']
E       [right]: [None]
```
On Windows at least fastparquet seems to name the index `index` even if there was no name for the index to start with.

dhirschfeld committed:
Fix test_write_with_index to work with fastparquet

fastparquet automatically names an index 'index' if it doesn't
already have a name

dhirschfeld commented:
The single Travis failure seems unrelated?

jreback commented:
linting issues, you can see locally by: ``make lint-diff``

```
The command "ci/script_multi.sh" exited with 0.
124.03s$ ci/lint.sh
inside ci/lint.sh
Linting *.py
pandas/io/parquet.py:6:1: F401 'pandas.compat.range' imported but unused
pandas/io/parquet.py:90:12: E121 continuation line under-indented for hanging indent
pandas/io/parquet.py:117:5: E303 too many blank lines (2)
pandas/io/parquet.py:118:23: E128 continuation line under-indented for visual indent
pandas/io/parquet.py:123:80: E501 line too long (83 > 79 characters)
pandas/io/parquet.py:153:80: E501 line too long (83 > 79 characters)
pandas/tests/io/test_parquet.py:261:1: W293 blank line contains whitespace
pandas/tests/io/test_parquet.py:281:67: E241 multiple spaces after ','
pandas/tests/io/test_parquet.py:302:80: E501 line too long (81 > 79 characters)
pandas/tests/io/test_parquet.py:305:34: E226 missing whitespace around arithmetic operator
Linting *.py DONE
Linting setup.py
```

dhirschfeld committed:
Fix linting issues

dhirschfeld committed:
Added wahtsnew entry for 0.21.1

dhirschfeld commented:
I added the whatsnew entry to 0.21.1 since the pyarrow version bump is no longer required.

If the version gets bumped in 0.22 then the crufty compatibility code can be removed then...

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into pr/18629

jorisvandenbossche committed:
move whatsnew note to parquet section

jorisvandenbossche commented on ci/requirements-2.7.sh:
```diff
@@ -4,4 +4,4 @@ source activate pandas
 echo "install 27"
 
-conda install -n pandas -c conda-forge feather-format pyarrow=0.4.1 fastparquet
+conda install -n pandas -c conda-forge feather-format pyarrow=0.7.0 fastparquet
```
let's keep the old version here? (so we actually test the older version, now we dropped bumping the version in this PR)

jorisvandenbossche committed:
undo version changes in ci and install docs

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -34,82 +35,155 @@ def get_engine(engine):
+        self.validate_dataframe(df)
+        if self._pyarrow_lt_070:
+            self._validate_write_lt_070(
+                df, path, compression, coerce_timestamps, **kwargs
```
Is there a reason you pass ` path, compression, coerce_timestamps` here?

dhirschfeld replied:
Just a left over from a refactoring. Should be removed as they're no longer needed...

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -34,82 +35,155 @@ def get_engine(engine):
-        if self._pyarrow_lt_060:
-            table = self.api.Table.from_pandas(df, timestamps_to_ms=True)
-            self.api.parquet.write_table(
-                table, path, compression=compression, **kwargs)
```
Do we need to keep this for older pyarrow versions ?

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -34,82 +35,155 @@ def get_engine(engine):
+        if not isinstance(df.index, Int64Index):
+            msg = (
+                "parquet does not support serializing {} for the index;"
+                "you can .reset_index() to make the index into column(s)"
```
maybe we should add here that they can also install latest pyarrow or fastparquet version (same for the others)

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -34,82 +35,155 @@ def get_engine(engine):
-class FastParquetImpl(object):
+        parquet_file = self.api.parquet.ParquetFile(path)
+        if self._pyarrow_lt_070:
+            parquet_file.path = path
```
I would just pass `path` to the `_read_lt_070` function?

jreback commented on pandas/tests/io/test_parquet.py:
```diff
@@ -6,6 +6,7 @@
 from warnings import catch_warnings
 
 import numpy as np
+from numpy.random import randn
```
we don't do the import like this anywhere else, appreciate changing it

jreback commented on pandas/tests/io/test_parquet.py:
```diff
@@ -323,6 +282,70 @@ def test_read_columns(self, engine):
+        df.columns = pd.MultiIndex.from_tuples([('a', 1), ('a', 2), ('b', 1)]),
+        self.check_error_on_write(df, engine, ValueError)
+
+    def test_multiindex_with_columns(self, engine):
```
you can just pass pa instead of engine to skip fp

jorisvandenbossche reviewed (commented):
@dhirschfeld Thanks for the update. I think there are still some left-overs to clean-up from the undo removing support for older versions, but I can also put some time in it this evening if needed.

jorisvandenbossche committed:
undo testing changes for old pyarrow versions

jorisvandenbossche committed:
fix test for older versions of pyarrow

jorisvandenbossche committed:
use read_pandas for pyarrow < 0.7

jorisvandenbossche committed:
further clean-up

jorisvandenbossche commented:
@dhirschfeld sorry for taking over your PR, but I want to get this merged for 0.21.1, and didn't know if you would have time today/tomorrow. 
I restored some of the testing code for testing the older versions, and fixed some bugs in the code to get those working. And addressed the comments of myself and @jreback.

But feel free to review my changes

dhirschfeld commented:
@jorisvandenbossche - that's fine, I'd only be able to look into it tonight so am happy for you to take the lead!  I'm hoping to make use of it so am keen to get it in too...

dhirschfeld commented:
LGTM!

jorisvandenbossche commented:
And all CI is passing!

@jreback more comments? 
(the compat for older versions is maybe not the cleanest, but it will be directly removed on master anyway .. (I think we want to do that?))

jreback commented on pandas/io/parquet.py:
```diff
@@ -34,39 +35,72 @@ def get_engine(engine):
+    def validate_dataframe(df):
+        if not isinstance(df, DataFrame):
+            raise ValueError("to_parquet only supports IO with DataFrames")
+        # must have value column names (strings only)
```
can you add a blank line before comments, easier to read

jreback reviewed (approved):
tiny style comments, otherwise lgtm.

jorisvandenbossche committed:
style update

jorisvandenbossche commented:
@dhirschfeld Thanks a lot for starting the PR!

dhirschfeld commented:
Thanks @jorisvandenbossche for getting this over the line! I'm excited to make use of this new functionality! :)
